File: src/graph.c
#include "graph.h"

#include "allocator.h"
#include "hash.h"
#include "hashmap.h"
#include "hashset.h"
#include "queue.h"

static u32
usize_hash(usize const u) {
    return fnv1a_u64(HASH_INIT, u);
}
static u32
usize_equals(usize const u1, usize const u2) {
    return u1 == u2;
}
HASHSET_IMPL(usize, sizes, usize_hash, usize_equals)
VECTOR_IMPL(struct group*, group_ptr)

static u32
edge_hash(struct edge edge) {
    return fnv1a_u64(fnv1a_u64(HASH_INIT, edge.from), edge.to);
}

static bool
edge_equals(struct edge edge1, struct edge edge2) {
    return edge1.from == edge2.from && edge1.to == edge2.to;
}

HASHSET_IMPL(struct edge, edge, edge_hash, edge_equals)
HASHMAP_IMPL(usize, struct hashset_sizes*, adjacency, usize_hash, usize_equals)

QUEUE_DEF(usize, sizes)
QUEUE_IMPL(usize, sizes)

static struct hashset_edge
graph_compute_transitive_edges(struct object_graph obj_graph) {
    struct hashset_edge transitive_edges = hashset_edge_copy(obj_graph.edges);
    hashmap_foreach(obj_graph.adjacency_lists, connector) {
        hashmap_foreach(obj_graph.adjacency_lists, from) {
            struct edge to_connector = {
                .from = from.key,
                .to   = connector.key,
            };
            hashmap_foreach(obj_graph.adjacency_lists, to) {
                struct edge full_jump = {
                    .from = from.key,
                    .to   = to.key,
                };

                if (hashset_edge_contains(transitive_edges, full_jump)) {
                    continue;
                }

                struct edge from_connector = {
                    .from = connector.key,
                    .to   = to.key,
                };

                if (hashset_edge_contains(transitive_edges, to_connector)
                    && hashset_edge_contains(
                        transitive_edges, from_connector
                    )) {
                    hashset_edge_add(&transitive_edges, full_jump);
                }
            }
        }
    }

    return transitive_edges;
}

HASHMAP_DEFS(usize, usize, group_ids)
HASHMAP_IMPL(usize, usize, group_ids, usize_hash, usize_equals)
HASHMAP_DEFS(usize, struct group_data*, group_data)
HASHMAP_IMPL(usize, struct group_data*, group_data, usize_hash, usize_equals)

static void
graph_create_groups(
    struct arena* arena, struct object_graph obj_graph,
    struct hashset_edge transitive_edges, struct hashmap_group_ids* group_ids,
    struct hashmap_group_data* group_data_map
) {
    usize id_counter = 0;

    hashmap_foreach(obj_graph.adjacency_lists, vertex) {
        if (hashmap_group_ids_contains(*group_ids, vertex.key)) {
            continue;
        }

        struct group_data* new_group
            = arena_allocate(arena, sizeof(struct group_data));
        new_group->functions      = hashset_sizes_new(arena);
        new_group->adjacency_list = hashset_sizes_new(arena);
        hashset_sizes_add(&new_group->functions, vertex.key);
        hashmap_group_data_add(group_data_map, id_counter, new_group);
        hashmap_group_ids_add(group_ids, vertex.key, id_counter);
        hashmap_foreach(obj_graph.adjacency_lists, other_vertex) {
            if (hashset_edge_contains(
                    transitive_edges,
                    (struct edge){ .from = vertex.key, .to = other_vertex.key }
                )
                && hashset_edge_contains(
                    transitive_edges, (struct edge){ .from = other_vertex.key,
                                                     .to   = other_vertex.key }
                )) {
                hashmap_group_ids_add(group_ids, other_vertex.key, id_counter);
                hashset_sizes_add(&new_group->functions, other_vertex.key);
            }
        }
        id_counter += 1;
    }
}

struct pair {
    u64 from;
    u64 to;
};

HASHSET_DEFS(struct pair, pair)
static u32
pair_hash(struct pair pair) {
    return fnv1a_u64(fnv1a_u64(HASH_INIT, pair.from), pair.to);
}
static bool
pair_equals(struct pair pair1, struct pair pair2) {
    return pair1.from == pair2.from && pair1.to == pair2.to;
}
HASHSET_IMPL(struct pair, pair, pair_hash, pair_equals)

static void
graph_create_edges(
    struct arena* arena, struct object_graph* obj_graph,
    struct hashmap_group_ids* group_ids,
    struct hashmap_group_data* group_data_map
) {
    struct hashset_pair group_edges = hashset_pair_new(arena);

    hashmap_foreach(obj_graph->adjacency_lists, vertex) {
        auto vertex_id   = hashmap_group_ids_find(group_ids, vertex.key);
        auto vertex_data = hashmap_group_data_find(group_data_map, vertex_id);
        hashset_foreach(*vertex.value, other_vertex) {
            auto other_id = hashmap_group_ids_find(group_ids, other_vertex);
            if (vertex_id == other_id) {
                continue;
            }
            if (hashset_pair_contains(
                    group_edges,
                    (struct pair){ .from = vertex_id, .to = other_id }
                )) {
                continue;
            }
            hashset_pair_add(
                &group_edges, (struct pair){ .from = vertex_id, .to = other_id }
            );
            hashset_sizes_add(&vertex_data->adjacency_list, other_id);
            hashmap_group_data_find(group_data_map, other_id)->indegree += 1;
        }
    }
}

static struct vector_group_ptr
graph_generate_order(
    struct arena* arena, struct hashmap_group_data* group_data_map
) {
    struct queue_sizes id_queue    = queue_sizes_new(arena);
    struct vector_group_ptr output = vector_group_ptr_new(arena);

    hashmap_foreach(*group_data_map, group) {
        if (group.value->indegree == 0) {
            queue_sizes_push(&id_queue, group.key);
        }
    }

    while (!queue_sizes_is_empty(id_queue)) {
        auto new_id     = queue_sizes_pop(&id_queue);
        auto group_data = hashmap_group_data_find(group_data_map, new_id);
        struct group* output_group
            = arena_allocate(arena, sizeof(struct group));
        output_group->members = hashset_sizes_copy(group_data->functions);

        hashset_foreach(group_data->adjacency_list, adjacenct_group) {
            auto data
                = hashmap_group_data_find(group_data_map, adjacenct_group);
            data->indegree -= 1;
            if (data->indegree == 0) {
                queue_sizes_push(&id_queue, adjacenct_group);
            }
        }

        vector_group_ptr_add(&output, output_group);
    }

    return output;
}

struct vector_group_ptr
graph_compute_order(struct arena* arena, struct object_graph* obj_graph) {
    struct hashset_edge transitive_edges
        = graph_compute_transitive_edges(*obj_graph);
    struct hashmap_group_ids group_ids       = hashmap_group_ids_new(arena);
    struct hashmap_group_data group_data_map = hashmap_group_data_new(arena);
    graph_create_groups(
        arena, *obj_graph, transitive_edges, &group_ids, &group_data_map
    );
    graph_create_edges(arena, obj_graph, &group_ids, &group_data_map);
    return graph_generate_order(arena, &group_data_map);
}

struct hashset_sizes*
graph_add_function(
    struct arena* arena, struct object_graph* obj_graph, usize const function
) {
    if (hashmap_adjacency_contains(obj_graph->adjacency_lists, function)) {
        return hashmap_adjacency_find(&obj_graph->adjacency_lists, function);
    } else {
        auto new_set                   = hashset_string_new(arena);
        struct hashset_sizes* heap_set = arena_allocate(arena, sizeof(new_set));
        memcpy(heap_set, &new_set, sizeof(new_set));
        hashmap_adjacency_add(&obj_graph->adjacency_lists, function, heap_set);
        return heap_set;
    }
}

void
graph_add_edge(
    struct arena* arena, struct object_graph* obj_graph, usize const from,
    usize const to
) {
    struct hashset_sizes* set = graph_add_function(arena, obj_graph, from);
    hashset_sizes_add(set, to);
    hashset_edge_add(
        &obj_graph->edges, (struct edge){ .from = from, .to = to }
    );
}

File: src/graph.zig
const std = @import("std");

const AstProperty = @import("ast.zig").AstProperty;
const TypecheckEnv = @import("types.zig").TypecheckEnv;

const Edge = struct {
    from: usize,
    to: usize,
};

const Group = struct {
    members: std.AutoHashMap(usize, void),
};

const GroupData = struct {
    functions: std.AutoHashMap(usize, void),
    adjacency_list: std.AutoHashMap(usize, void),
    indegree: usize,

    fn init(allocator: std.mem.Allocator) !*GroupData {
        const group_data = try allocator.create(GroupData);
        group_data.* = .{
            .functions = std.AutoHashMap(usize, void).init(allocator),
            .adjacency_list = std.AutoHashMap(usize, void).init(allocator),
            .indegree = 0,
        };
        return group_data;
    }
};

pub const ObjectGraph = struct {
    allocator: std.mem.Allocator,
    adjacency_list: std.AutoHashMap(usize, std.AutoHashMap(usize, void)),
    edges: std.AutoHashMap(Edge, void),

    pub fn init(allocator: std.mem.Allocator) @This() {
        return .{
            .allocator = allocator,
            .adjacency_list = std.AutoHashMap(usize, std.AutoHashMap(usize, void)).init(allocator),
            .edges = std.AutoHashMap(Edge, void).init(allocator),
        };
    }

    pub fn addFunction(self: *@This(), from: usize) !*std.AutoHashMap(usize, void) {
        const search_result = self.adjacency_list.getPtr(from);
        if (search_result) |value| {
            return value;
        } else {
            const new_set = std.AutoHashMap(usize, void).init(self.allocator);
            try self.adjacency_list.put(from, new_set);
            return self.adjacency_list.getPtr(from).?;
        }
    }

    pub fn initProperties(self: *@This(), properties: std.ArrayList(*AstProperty), env: *TypecheckEnv) !void {
        for (properties.items) |prop| {
            try env.addName(prop.name, prop.id);
        }

        for (properties.items) |prop| {
            _ = try self.addFunction(prop.id);
            switch (prop.value.value) {
                .properties => |props| {
                    for (props.items) |p| {
                        try self.addEdge(p.id, prop.id);
                    }
                    var new_env = TypecheckEnv.init(env.allocator, env);
                    try self.initProperties(prop.value.value.properties, &new_env);
                },
                .property_access => |ids| {
                    const name = ids.items[0];
                    const id = env.find(name);
                    try self.addEdge(id, prop.id);
                },
                else => {},
            }
        }
    }

    pub fn addEdge(self: *@This(), from: usize, to: usize) !void {
        var set = try self.addFunction(from);
        try set.put(to, {});
        try self.edges.put(.{
            .from = from,
            .to = to,
        }, {});
    }

    pub fn computeOrder(self: *@This()) !std.ArrayList(*Group) {
        const transitive_edges = try self.computeTransitiveEdges();
        var group_ids = std.AutoHashMap(usize, usize).init(self.allocator);
        var group_data_map = std.AutoHashMap(usize, *GroupData).init(self.allocator);
        try self.createGroups(transitive_edges, &group_ids, &group_data_map);
        try self.createEdges(&group_ids, &group_data_map);
        return try self.generateOrder(group_data_map);
    }

    fn computeTransitiveEdges(self: *@This()) !std.AutoHashMap(Edge, void) {
        var transitive_edges = try self.edges.clone();
        var it_connector = self.adjacency_list.keyIterator();
        while (it_connector.next()) |connector| {
            var it_from = self.adjacency_list.keyIterator();
            while (it_from.next()) |from| {
                const to_connector: Edge = .{
                    .from = from.*,
                    .to = connector.*,
                };
                var it_to = self.adjacency_list.keyIterator();
                while (it_to.next()) |to| {
                    const full_jump: Edge = .{
                        .from = from.*,
                        .to = to.*,
                    };
                    if (transitive_edges.contains(full_jump)) {
                        continue;
                    }
                    const from_connector: Edge = .{
                        .from = connector.*,
                        .to = to.*,
                    };
                    if (transitive_edges.contains(to_connector) and transitive_edges.contains(from_connector)) {
                        try transitive_edges.put(full_jump, {});
                    }
                }
            }
        }
        return transitive_edges;
    }

    fn createGroups(self: @This(), transitive_edges: std.AutoHashMap(Edge, void), group_ids: *std.AutoHashMap(usize, usize), group_data_map: *std.AutoHashMap(usize, *GroupData)) !void {
        var id_counter: usize = 0;
        var it_vertex = self.adjacency_list.keyIterator();
        while (it_vertex.next()) |vertex| {
            if (group_ids.contains(vertex.*)) {
                continue;
            }

            var new_group = try GroupData.init(self.allocator);
            try new_group.functions.put(vertex.*, {});
            try group_data_map.put(id_counter, new_group);
            try group_ids.put(vertex.*, id_counter);
            var it_other_vertex = self.adjacency_list.keyIterator();
            while (it_other_vertex.next()) |other_vertex| {
                if (transitive_edges.contains(.{ .from = vertex.*, .to = other_vertex.* }) and transitive_edges.contains(.{ .from = other_vertex.*, .to = vertex.* })) {
                    try group_ids.put(other_vertex.*, id_counter);
                    try new_group.functions.put(other_vertex.*, {});
                }
            }

            id_counter += 1;
        }
    }

    fn createEdges(self: *@This(), group_ids: *std.AutoHashMap(usize, usize), group_data_map: *std.AutoHashMap(usize, *GroupData)) !void {
        var group_edges = std.AutoHashMap(Edge, void).init(self.allocator);
        var it_vertex = self.adjacency_list.iterator();
        while (it_vertex.next()) |vertex| {
            const vertex_id = try group_ids.getOrPut(vertex.key_ptr.*);
            const vertex_data = try group_data_map.getOrPut(vertex_id.value_ptr.*);
            var it_other_vertex = vertex.value_ptr.*.keyIterator();
            while (it_other_vertex.next()) |other_vertex| {
                const other_id = try group_ids.getOrPut(other_vertex.*);
                if (other_id.value_ptr.* == vertex_id.value_ptr.*) {
                    continue;
                }
                if (group_edges.contains(.{ .from = vertex_id.value_ptr.*, .to = other_id.value_ptr.* })) {
                    continue;
                }
                try group_edges.put(.{
                    .from = vertex_id.value_ptr.*,
                    .to = other_id.value_ptr.*,
                }, {});
                try vertex_data.value_ptr.*.adjacency_list.put(other_id.value_ptr.*, {});
                group_data_map.get(other_id.value_ptr.*).?.indegree += 1;
            }
        }
    }

    fn generateOrder(self: *@This(), group_data_map: std.AutoHashMap(usize, *GroupData)) !std.ArrayList(*Group) {
        var id_queue = std.fifo.LinearFifo(usize, .Dynamic).init(self.allocator);
        var output = std.ArrayList(*Group).init(self.allocator);

        var it = group_data_map.iterator();
        while (it.next()) |kv| {
            if (kv.value_ptr.*.indegree == 0) {
                try id_queue.writeItem(kv.key_ptr.*);
            }
        }

        while (id_queue.readItem()) |new_id| {
            const group_data = group_data_map.get(new_id).?;
            var output_group = try self.allocator.create(Group);
            output_group.members = try group_data.functions.clone();

            var group_it = group_data.adjacency_list.keyIterator();
            while (group_it.next()) |adjacent_group| {
                const data = group_data_map.get(adjacent_group.*).?;
                data.indegree -= 1;
                if (data.indegree == 0) {
                    try id_queue.writeItem(adjacent_group.*);
                }
            }

            try output.append(output_group);
        }

        return output;
    }
};
